proposals/20210817-firebase_ml_publisher_component.md

```diff
@@ -0,0 +1,45 @@
+#### SIG TFX-Addons
+# Project Proposal
+
+**Your name:** Chansung Park
+
+**Your email:** deep.diver.csp@gmail.com
+
+**Your company/organization:** Individual(ML GDE)
+
+**Project name:** [Firebase ML Publisher](https://github.com/tensorflow/tfx-addons/issues/59)
+
+## Project Description
+This project defines a custom TFX component to publish/update ML models from TFX Pusher to [Firebase ML](https://firebase.google.com/products/ml). The input model from TFX Pusher is assumed to be a TFLite format.
+
+## Project Category
+Component
+
+## Project Use-Case(s)
+This project helps users to publish trained models directly from TFX Pusher component to Firebase ML. 
+
+With Firebase ML, we can guarantee that mobile devices can be equipped with the latest ML model without explicitly embedding binary in the project compiling stage. We can even A/B test different versions of a model with Google Analytics when the model is published on Firebase ML.
+
+## Project Implementation
+Firebase ML Publisher component will be implemented as Python function-based component. You can find the [actual source code](https://github.com/sayakpaul/Dual-Deployments-on-Vertex-AI/blob/main/custom_components/firebase_publisher.py) in my personal project. 
+
+The implementation details
+- Define a custom Python function-based TFX component. It takes the following parameters from a previous component.
+  - The URI of the pushed model from TFX Pusher component.
+  - Requirements from Firebase ML (credential JSON file path, Firebase temporary-use GCS bucket). Please find more information from [Before you begin section](https://firebase.google.com/docs/ml/manage-hosted-models#before_you_begin) in the official Firebase document.
```

Comment:
I am not sure how I can do that.
Since every component runs on a docker container, the environment variable should be set in Dockerfile then? (please let me know how to setup ENV variable for TFX component).

but still, we can make this parameter Optional. What do you think?

Reply:
When using Kubernetes on GKE, you can use Workload identity to set up your pods to automatically auth  as the user (see: https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity)

Otherwise, you can set it up using https://github.com/tensorflow/tfx/blob/master/tfx/orchestration/kubeflow/kubeflow_dag_runner.py#L78 (which is preferred over setting a GCS path)

Reply:
@casassg 

I am currently working on implementing this component, but I can't get it working correctly in Vertex AI Pipeline environment. 

Just like your comments, I thought Default Application Credential should work OK, but I get the error message like below by complaining to include `private key` as the credential.

I gave appropriate permissions to the Service Account which is `compute` such as `Firebase Admin` (it is not permission error anyways). I have tried out the same thing step-by-step in Vertex AI Workbench environment as well, and I get the same error. 

Do you have any clue? 

```
AttributeError: you need a private key to sign credentials.the credentials you are currently using <class 'google.auth.compute_engine.credentials.Credentials'> just contains a token. see https://googleapis.dev/python/google-api-core/latest/auth.html#setting-up-a-service-account for more details.
```

Reply:
I think we can figure this out as we merge the implementation as it's hard to debug without the code